activation/activation.go

```diff
@@ -467,36 +478,57 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
 		log.Uint32("view_cnt", activeSetSize))
 
-	signedAtx, err := b.SignAtx(atx)
-	if err != nil {
+	atxReceived := make(chan struct{})
+	b.atxPool.SetListener(func(atxHeader *types.ActivationTxHeader) {
+		if atxHeader.Id() == atx.Id() {
+			close(atxReceived)
+			b.atxPool.SetListener(nil)
+		}
+	})
+
+	if err := b.signAndBroadcast(atx); err != nil {
 		return err
 	}
 
-	buf, err := types.InterfaceToBytes(signedAtx)
-	if err != nil {
-		return err
+	fields := []log.Field{
+		log.AtxId(atx.ShortString()),
+		log.String("prev_atx_id", atx.PrevATXId.ShortString()),
+		log.String("post_atx_id", atx.PositioningAtx.ShortString()),
+		log.LayerId(uint64(atx.PubLayerIdx)),
+		log.EpochId(uint64(atx.PubLayerIdx.GetEpoch(b.layersPerEpoch))),
+		log.Uint32("active_set", atx.ActiveSetSize),
+		log.String("miner", b.nodeId.Key[:5]),
+		log.Int("view", len(atx.View)),
+	}
+	select {
+	case <-atxReceived:
+		b.log.Event().Info("atx published!", fields...)
+	case <-time.After(b.broadcastTimeout):
+		b.log.With().Error("broadcast timeout when publishing ATX", fields...)
```

Comment:
should we discard the current challenge?

Reply:
I don't think. Most likely the failure was due to the previous ATX not being in the database yet (waiting for miners to include it and pbase to advance). By leaving the state we'll retry every layer.

Once we fix #1552, there will also be a termination condition to the retries.

Reply:
if prevAtx was never included in the mesh this will never terminate until an epoch will pass

Reply:
if prevAtx was never published and will never be on the mesh, an ATX will not be built in this epoch and only after state is discarded a new atx will be published. do you think we should handle this?

Reply:
You're right, we won't publish an ATX in the current epoch, but how can we tell that the previous ATX will **never** be published?

I think that this can only be fixed when we support free-standing ATXs. Until then there's always a risk of blacklisting yourself if you discard state.

@barakshani thoughts?

Reply:
1. no need for pbase to advance - a block (and its content) is in the database once received (and syn. valid). Anyway, with free-standing that should change.
2. why a new atx will not be built/published if the previous is not on the mesh? is it an outcome of this PR's new code?

Reply:
ok, so you meant this timeout condition. with this new code we never clear b.nipst, hence we will not try to build a new atx.
What would actually happen then? At the beginning of a layer, we enter the function and go to line 440. Would we try to re-broadcast the atx? This should be the desirable behaviour, along with issue #1552.
In my opinion, it is not super critical for testnet, and if the miner has not received the atx until the timeout, we can assume it is lost forever, and start building a new one, with _previous atx_ the last one actually published.

Reply:
Cool, so this is what happens with this code.

Each layer we'll try to broadcast, and if it fails, just try again next layer. The only reason it should fail is due to the previous ATX not being in the ATXDB yet, which only currently happens when pbase passes a layer and the block is contextually valid.

Currently, this will happen forever. Once we implement #1552 we'll only try until the ATX is no longer relevant and then assume it's lost forever.

Once free-standing ATXs are implemented (#1059) this should really never happen since ATXs will be in the DB immediately, and we'll validate that an ATX hits the DB instead of the mempool, so the previous one will definitely be there already.

Reply:
Cool, so this is what happens with this code.

Each layer we'll try to broadcast, and if it fails, just try again next layer. The only reason it should fail is due to the previous ATX not being in the ATXDB yet, which only currently happens when pbase passes a layer and the block is contextually valid.

Currently, this will happen forever. Once we implement #1552 we'll only try until the ATX is no longer relevant and then assume it's lost forever.

Once free-standing ATXs are implemented (#1059) this should really never happen since ATXs will be in the DB immediately, and we'll validate that an ATX hits the DB instead of the mempool, so the previous one will definitely be there already.